sdk/nodejs/appservice/zMixins.ts

```diff
@@ -246,81 +255,257 @@ export interface HostSettings {
-            map[name + "/" + path] = value;
-        }
+    /**
+     * Serialized function callback.
+     */
+    body: Promise<pulumi.runtime.SerializedFunction>;
+
+    /**
+     * Application settings required by the function.
+     */
+    appSettings?: pulumi.Input<{ [key: string]: any }>;
 
-        return map;
-    });
+    /**
+     * HTTP route to call this function (applies to functions reachable via HTTP).
+     */
+    route?: pulumi.Input<string>;
 }
 
-function redirectConsoleOutput<C extends Context<R>, E, R extends Result>(callback: Callback<C, E, R>) {
-    return (context: C, event: E) => {
-        // Redirect console logging to context logging.
-        console.log = context.log;
-        console.error = context.log.error;
-        console.warn = context.log.warn;
-        // tslint:disable-next-line:no-console
-        console.info = context.log.info;
+async function produceDeploymentPackage(args: MultiFunctionAppArgs): Promise<pulumi.asset.AssetMap> {
+    
+    if (args.functions === undefined) {
+        throw new Error("Can not serialize undefined Azure Functions");
+    }
 
-        return callback(context, event);
-    };
+    const map: pulumi.asset.AssetMap = {};
+    map["host.json"] = new pulumi.asset.StringAsset(JSON.stringify({
+        version: "2.0",
+        tracing: { consoleLevel: "verbose" },
+        extensionBundle: {
+            id: "Microsoft.Azure.Functions.ExtensionBundle",
+            version: "[1.*, 2.0.0)"
+        },
+        ...args.hostSettings,
+    }));
+
+    const pathSet = await pulumi.runtime.computeCodePaths(args.codePathOptions);
+    for (const [path, value] of pathSet.entries()) {
+        map[path] = value;
+    }
+
+    for (const func of args.functions) {
+        map[`${func.name}/function.json`] = pulumi.output(func.bindings).apply(bs => new pulumi.asset.StringAsset(JSON.stringify({
+            disabled: false,
+            bindings: bs,
+        })));
+
+        const body = await func.body;
+
+        map[`${func.name}/index.js`] = new pulumi.asset.StringAsset(`module.exports = require("./handler").handler`),
+        map[`${func.name}/handler.js`] = new pulumi.asset.StringAsset(body.text);
+    }
+
+    return map;
 }
 
+function combineAppSettings(args: MultiFunctionAppArgs): pulumi.Output<{[key: string]: string}> {
+    const applicationSetting = args.appSettings || {};
+    const perFunctionSettings = args.functions !== undefined ? args.functions.map(c => c.appSettings || {}) : [];
+    return pulumi.all([applicationSetting, ...perFunctionSettings]).apply(items => items.reduce((a, b) => ({ ...a, ...b }), {}));
+}
+
+export type MultiFunctionAppArgs = util.Overwrite<FunctionAppArgs, {
+
+    /**
+     * The functions to deploy as parts of this application.
+     */
+    functions?: FunctionArgs[];
+
+    /**
+     * The deployment package to deploy as-is (instead of the 'functions' property).
+     */
+    archive?: pulumi.Input<pulumi.asset.Archive>;

```

Comment:
consider explicitly doc'ing the relationship between props.  i.e.: Either one of [functions] or [archive] must be provided.